asf_welsh_energy_consultation/pipeline/process_data.py

```diff
@@ -384,3 +384,122 @@ def generate_age_data(wales_df):
+
+    # Sort by date ascending
+    installations_by_year = installations_by_year.sort_values("year")
+    installations_by_year = installations_by_year.rename(
+        columns={"commission_date": "date"}
+    )
+
+    return installations_by_year
+
+
+def mean_installations_per_year(min_year, max_year):
+    """
+    Get mean average MCS installations in Wales per year for given date range.
+    Args:
+        min_year: Minimum year (exclusive)
+        max_year: Maximum year (exclusive)
+
+    Returns:
+        int: Mean average MCS installations per year.
+    """
+    installations_by_year = get_installations_per_year()
+    subset = installations_by_year[
+        (installations_by_year["year"] > min_year)
+        & (installations_by_year["year"] < max_year)
+    ]
+
+    return subset["n"].mean()
+
+
+def median_installations_per_year(min_year, max_year):
+    """
+    Get median average MCS installations in Wales per year for given date range.
+    Args:
+        min_year: Minimum year (exclusive)
+        max_year: Maximum year (exclusive)
+
+    Returns:
+        int: Median average MCS installations per year.
+    """
+    installations_by_year = get_installations_per_year()
+    subset = installations_by_year[
+        (installations_by_year["year"] > min_year)
+        & (installations_by_year["year"] < max_year)
+    ]
+
+    return subset["n"].median()
+
+
+def get_total_rural_and_urban_properties():
+    """
+    Get total count of properties in Wales in urban vs rural locations.
+
+    Returns:
+        dict: Percent of rural and urban properties in Wales.
+    """
+    rural = get_data.get_rurality()
+    dwellings = get_data.get_dwelling_data()
+
+    df = dwellings.merge(rural, how="left", on="lsoa_code")
+    df = df[df["country"] == "W"]
+    rurality = df.groupby("rural_2")["total_dwellings"].sum().to_dict()
+    rurality_pct_dict = {
+        "Rural": (rurality["Rural"] / (rurality["Rural"] + rurality["Urban"])) * 100,
+        "Urban": (rurality["Urban"] / (rurality["Rural"] + rurality["Urban"])) * 100,
+    }
+
+    return rurality_pct_dict
+
+
+def get_total_on_off_gas_postcodes():
+    """
+    Get total count of postcodes in Wales which are on- vs off-gas.
+
+    Returns:
+        dict: Percent of on- and off-gas postcodes in Wales.
+    """
+
+    og = get_data.get_offgas()
+    postcodes = get_data.get_countries()
+
+    postcodes_og = postcodes.merge(og, how="left", on="postcode")
+
+    # Any postcodes not in off gas dataset have NA values in 'off_gas' col
+    # We assume the remaining postcodes are on gas and fillna with 'on gas'
+    postcodes_og["off_gas"] = postcodes_og["off_gas"].replace({True: "Off gas"})
+    postcodes_og["off_gas"] = postcodes_og["off_gas"].fillna("On gas")
+
+    # Filter for Wales only
+    wales_og = postcodes_og.loc[postcodes_og["country"] == "Wales"].copy()
+
+    # Calculate % of postcodes on and off gas
+    wales_og_dict = wales_og["off_gas"].value_counts(dropna=False).to_dict()
```

Comment:
does it matter if we set dropna to True or False? It shoudln't right? Because "off_gas" should never be missing after we do `postcodes_og["off_gas"] = postcodes_og["off_gas"].fillna("On gas")`

Reply:
It shouldn't matter because we filled, but `dropna=False` serves as an additional check that the `fillna` has been successful and worked as expected.